### PR TITLE
[4.0] Restores the possibility to show rule descriptions in rules formfield.

### PR DESCRIPTION
**Pull request by Bakual**
With https://github.com/joomla/joomla-cms/pull/19285, the possibility to add a description to a custom rule was removed.
@wilsonge reverted part of the change so it should still be possible to set a description, ~but he made a typo in his revert~ :smile: 

### Summary of Changes
This PR fully restores the possibility to set and display a description in the rules field.
It incorporates a check for the presence of the description so the rules that no longer have a description don't show a tooltip.

### Testing Instructions
* Make sure the rules in core extensions don't show any description. They should all be removed with the original PR and still don't show any description.
* Test with a 3rd party extension (eg https://github.com/Bakual/SermonSpeaker/releases/download/6.0.0-alpha1/com_sermonspeaker.zip) that rules with a description still show a tooltip OR add a description to the rules in access.xml file of com_content and test if it shows up.

### Expected result
A description is shown
Please note that the tooltip itself is broken due to Bootstrap tooltips not being loaded on that page. The method used in core is `JHtml::tooltipText()` which doesn't load Bootstrap by itself. Same issue exists in other places (eg extension manager) as well and needs to be fixed independant of this PR.
The description will show as a native "title".

### Actual result
Nothing is shown

### Documentation Changes Required
None

### CODE REVIEW

**Comment by ggppdk** [comment]
I ll test later today

This PR is useful , because with it we do not have the requirement that all 3rd parties
- create a layout override just for adding the tooltip !

... and then breaking if the CORE layout is updated with something important
and also it is benefitial to the CORE, at any time if it is decided that an existing or new ACL rule needs a description (tooltip), then only the XML file will need to be updated

**Comment by wilsonge** [comment]
I've manually patched the typo direct to 4.0 (thanks!!) so we can focus on the layout changes here :)

File: libraries/src/Form/Field/RulesField.php [hunk 1]
```diff
@@ -302,7 +302,9 @@ protected function getInput()
 			{
 				$html[] = '<tr>';
 				$html[] = '<td headers="actions-th' . $group->value . '">';
-				$html[] = '<label for="' . $this->id . '_' . $action->name . '_' . $group->value . '">';
+				$description = (!empty($action->description)) ?  ' class="hasTooltip" title="'
```
Comment (Quy) [inline libraries/src/Form/Field/RulesField.php]: Remove extra space afte `?`

**Comment by dgrammatiko** [comment]
Just couple notes here: 
- J4 will **not** use the bootstrap tooltips (meaning the class hasTooltip will be useless)
- J4 will have a new tooltip (meaning the HTML and the CSS and possibly the JS will be totally different , eg either only CSS or Custom Element).

So one way or another this layout (with or without the description) will not be B/C to 3.x ones.

**Comment by Bakual** [comment]
> So one way or another this layout (with or without the description) will not be B/C to 3.x ones.

Of course. Along with a lot of other layouts that need to be adjusted to the new way once that work is done.

**Comment by ggppdk** [comment]
I have tested this item :white_check_mark: successfully on 45f1c77b9c5d1c71988462567427829ddd701dcc<br /><br />Tooltip is shown when description is added to the XML
when description is not present in XML, then no tooltip shown

About the new tooltip JS, yes i understand that 'hasTooltip' class will be found and replaced when new tooltips are added
now it has no special effect, and the tooltip is shown as browser tooltip after 1s of hovering over the rule label<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19536">issues.joomla.org/tracker/joomla-cms/19536</a>.</sub>

**Comment by wilsonge** [comment]
Sorry I merged the PR that converted the form field to layouts. Going to need an update here

**Comment by Bakual** [comment]
No problem. Will look at it later today.
For my reference: #19507

**Comment by Bakual** [comment]
Redid the PR for the new JLayout.

**Comment by Bakual** [comment]
Please note: After checking with @Anu1601CS I also move the JLayout to a more proper place.
Name from `joomla/form/field/rules/tabs.php` to `joomla/form/field/rules.php`.

**Comment by anuragteapot** [comment]
@Bakual  Thanks,
Should we also need tooltip in <code>select</code> option. ?

**Comment by Bakual** [comment]
> Should we also need tooltip in select option. ?

The way it is now is the same as it was before PR https://github.com/joomla/joomla-cms/pull/19285.
I think that was sufficient.

**Comment by anuragteapot** [comment]
Ok, also we need to add language files

**Comment by Bakual** [comment]
> Ok, also we need to add language files

Why? Core doesn't use descriptions anymore, thus no language strings. If 3rd parties want to show a description, they need to add them of course.

**Comment by dgrammatiko** [comment]
@Bakual can you also correct the doc blocks in both files, they're a bit off...

**Comment by anuragteapot** [comment]
Ok, this change only for 3rd parties if they want to show a description.
That's why this PR.
Thanks

**Comment by Bakual** [comment]
> can you also correct the doc blocks in both files, they're a bit off...

I'm not sure what you mean. Anyway I didn't touch them and I don't want to stuff more into this PR than needed.

**Comment by anuragteapot** [comment]
@Bakual I am going to correct the doc block variable.
Should I move <code>joomla/form/field/rules/tabs.php</code> to <code>joomla/form/field/rules.php</code> ? In my PR

**Comment by Bakual** [comment]
You can fix it in the original place. Gives less conflicts. Git will figure out the file moving itself.

**Comment by infograf768** [comment]
Branch out of date

**Comment by Bakual** [comment]
It's out of date as soon as something is merged into 4.0-dev. However it doesn't have any conflicts, right?
So this should be still testable without me merging in 4.0 every day :smile:

**Comment by brianteeman** [comment]
@bakual that is correct

**Comment by laoneo** [comment]
Can we get some retests here to make sure we do not break anything. Thanks.

**Comment by infograf768** [comment]
I guess there was a change in RulesField since this PR was proposed which breaks the permissions display

```
@@ -39,7 +39,7 @@ class RulesField extends FormField
 	 * @var    string
 	 * @since  __DEPLOY_VERSION__
 	 */
-	protected $layout = 'joomla.form.field.rules.tabs';
+	protected $layout = 'joomla.form.field.rules';
 
```
should be taken off from this PR. We need the `joomla.form.field.rules.tabs` layout to display the permissions.

We have here another issue which is the way the TIP is displayed.
To test, I re-added a description constant for com_content.

screen shot 2018-04-04 at 11 08 24

**Comment by infograf768** [comment]
@Bakual

**Comment by Bakual** [comment]
> We have here another issue which is the way the TIP is displayed.

That is expected as I wrote in the PR description under "Expected result"

> Please note that the tooltip itself is broken due to Bootstrap tooltips not being loaded on that page. The method used in core is JHtml::tooltipText() which doesn't load Bootstrap by itself. Same issue exists in other places (eg extension manager) as well and needs to be fixed independant of this PR.
The description will show as a native "title".

> I guess there was a change in RulesField since this PR was proposed which breaks the permissions display

See: https://github.com/joomla/joomla-cms/pull/19536#issuecomment-363077792
> Please note: After checking with @Anu1601CS I also move the JLayout to a more proper place.
Name from joomla/form/field/rules/tabs.php to joomla/form/field/rules.php.

Should be all as expected.

**Comment by infograf768** [comment]
if i patch with this pr, the permissions tab does display nothing. will try with a fresh install

**Comment by Bakual** [comment]
Maybe the patchtester can't deal correctly with file movings?

**Comment by infograf768** [comment]
@Bakual 
I never use patchtester . :)

OK. I now patched after updating the layout folder.

All works fine here. This should be merged now before we have more changes . 🥂

**Pull request merged**